common/src/math.rs

```diff
@@ -1,34 +1,107 @@
 use serde::{Deserialize, Serialize};
 use std::ops::*;
 
+/// A stack-allocated 4-dimensional column-vector in Minkowski space. Such
+/// vectors are useful for computations in the hyperboloid model of hyperbolic
+/// space. Note that the last coordinate, not the first coordinate, is treated
+/// as the special "time" coordinate.
+///
+/// This vector type is versatile, being able to represent multiple things in
+/// Hyperbolic space. What it can represent is generally determined by the
+/// Minkowski inner product between the vector and itself.
+/// - If it's negative, it represents a point in hyperbolic space. The origin is
+///   represented with the unit w-vector.
+/// - If it's zero, it represents an _ideal_ point in hyperbolic space. Such a
+///   point can be associated with horospheres
+/// - If it's positive, it represents an _ultraideal_ point in hyperbolic space.
+///   Such points can be treated as oriented planes.
+///
+/// If the absolute value of this Minkowski inner product is 1, it is
+/// normalized, and equations involving such a vector tend to be simpler, much
+/// like with unit vectors.
+///
+/// Note that the simplest way to represent directions/velocities/normals at a
+/// point in hyperbolic space is with a vector whose Minkowski inner product
+/// with that point is 0. Such a vector will be tangent to the hyperboloid model
+/// at that associated point, so it can naturally represent movement along the
+/// hyperboloid in that direction.
+///
+/// As a general rule, when working with such vectors, it is highly recommended
+/// to avoid dot products and related operations such as vector magnitude, as
+/// these operations are meaningless in Minkowski space and are not preserved by
+/// isometries.
 #[derive(Debug, Copy, Clone, Serialize, Deserialize, PartialEq)]
 #[repr(C)]
 pub struct MVector<N: Scalar>(na::Vector4<N>);
 
+/// An `MVector` with the constraint that the Minkowski inner product between
+/// the vector and itself is -1. Such a vector can be used to represent a point
+/// in hyperbolic space.
 #[derive(Debug, Copy, Clone, Serialize, Deserialize, PartialEq)]
-#[repr(C)]
-pub struct MIsometry<N: Scalar>(na::Matrix4<N>);
+#[repr(transparent)]
+pub struct MUnitPointVector<N: Scalar>(MVector<N>);
+
+impl<N: Scalar> Deref for MUnitPointVector<N> {
+    type Target = MVector<N>;
 
-impl<N: RealField> From<na::Unit<na::Vector3<N>>> for MVector<N> {
-    fn from(value: na::Unit<na::Vector3<N>>) -> Self {
-        Self(value.into_inner().push(na::zero()))
+    fn deref(&self) -> &Self::Target {
+        &self.0
     }
 }
 
+/// An `MVector` with the constraint that the Minkowski inner product between
+/// the vector and itself is 1. Such a vector can be used to represent a
+/// direction in hyperbolic space.
+#[derive(Debug, Copy, Clone, Serialize, Deserialize, PartialEq)]
+#[repr(transparent)]
+pub struct MUnitDirectionVector<N: Scalar>(MVector<N>);
+
+impl<N: Scalar> Deref for MUnitDirectionVector<N> {
```

Comment:
Avoid unnecessary trait bounds, here and below.

Reply:
There is a compiler error if I remove the `Scalar` trait bound. The reason for this is most likely due to the bound present in `MVector`:
```rust
#[derive(Debug, Copy, Clone, Serialize, Deserialize, PartialEq)]
pub struct MVector<N: Scalar>(na::Vector4<N>);
```

That trait bound is in turn needed for the derived implementation of `Serialize` and `Deserialize` to work on `na::Vector4`.

I don't consider this to be too restrictive because of internal nalgebra code:
```rust
/// The basic scalar type for all structures of `nalgebra`.
///
/// This does not make any assumption on the algebraic properties of `Self`.
pub trait Scalar: 'static + Clone + PartialEq + Debug {}

impl<T: 'static + Clone + PartialEq + Debug> Scalar for T {}
```

Reply:
Resolving based on https://github.com/Ralith/hypermine/pull/446#issuecomment-2664547088